src/core/VectorGraphModel.cpp

```diff
@@ -0,0 +1,2411 @@
+		}
+	}
+	return output;
+}
+
+void VectorGraphDataArray::setDataArray(std::vector<PointF>* inputDataArray,
+	bool shouldCurve, bool shouldClear, bool shouldClamp, bool shouldRescale, bool shouldSort, bool callDataChanged)
+{
+#ifdef VECTORGRAPH_DEBUG_PAINT_EVENT
+	qDebug("setDataArray start");
+#endif
+	if (shouldClear == true)
+	{
+		m_dataArray.clear();
+	}
+	m_dataArray.resize(inputDataArray->size());
+	if (m_dataArray.size() == 0)
+	{
+		clearedEvent();
+	}
+	if (shouldClamp == true || shouldRescale == true || shouldSort == true)
+	{
+		formatArray(inputDataArray, shouldClamp, shouldRescale, shouldSort, false);
+	}
+	bool noneBefore = true;
+	bool isNegativeBefore = false;
+	for (unsigned int i = 0; i < m_dataArray.size(); i++)
+	{
+		m_dataArray[i].m_x = (*inputDataArray)[i].first;
+		m_dataArray[i].m_y = (*inputDataArray)[i].second;
+		// calculating curves
+		if (shouldCurve == true && i > 0)
+		{
+			// TODO take in to account x coords
+			float diff = m_dataArray[i - 1].m_y - m_dataArray[i].m_y;
+			// if before is bigger than after
+			bool isNegative = diff >= 0;
+			float direction = 0;
+			if (noneBefore == true)
+			{
+				direction = isNegative == true ? 1.0f : -1.0f;
+			}
+			else
+			{
+				direction = isNegative == isNegativeBefore ? 1.0f : isNegative == true ? -1.0f : 1.0f;
+			}
+
+			diff = diff * diff * 10.0f * direction;
+			diff = std::clamp(diff, -0.7f, 0.7f);
+			m_dataArray[i - 1].m_c = diff;
+
+			noneBefore = diff < 0.1f && diff > -0.1f;
+			isNegativeBefore = isNegative;
+		}
+	}
+	// the whole m_dataArray needs to be updated
+	getUpdatingFromPoint(-1);
+	if (callDataChanged == true)
+	{
+		dataChanged();
+	}
+}
+void VectorGraphDataArray::setDataArray(std::vector<float>* inputDataArray,
+	bool shouldCurve, bool shouldClear, bool shouldClamp, bool shouldRescale, bool callDataChanged)
+{
+	std::vector<PointF> convertedDataArray(inputDataArray->size());
+	float stepSize = 1.0f / static_cast<float>(convertedDataArray.size());
+	for (unsigned int i = 0; i < inputDataArray->size(); i++)
+	{
+		convertedDataArray[i].first = i * stepSize;
+		convertedDataArray[i].second = (*inputDataArray)[i];
+	}
+	setDataArray(&convertedDataArray, shouldCurve, shouldClear, shouldClamp, shouldRescale, false, callDataChanged);
+}
+void VectorGraphDataArray::setDataArray(float* inputDataArray, unsigned int size,
+	bool shouldCurve, bool shouldClear, bool shouldClamp, bool shouldRescale, bool callDataChanged)
+{
+	std::vector<PointF> convertedDataArray(size);
+	float stepSize = 1.0f / static_cast<float>(size);
+	for (unsigned int i = 0; i < size; i++)
+	{
+		convertedDataArray[i].first = i * stepSize;
+		convertedDataArray[i].second = inputDataArray[i];
+	}
+	setDataArray(&convertedDataArray, shouldCurve, shouldClear, shouldClamp, shouldRescale, false, callDataChanged);
+}
+
+unsigned int VectorGraphDataArray::setX(unsigned int pointLocation, float newX)
+{
+	int location = pointLocation;
+	if (m_isFixedX == false && newX <= 1.0f)
```

Comment:
would make sense to invert this too.
```suggestion
	if (m_isFixedX == true || newX > 1.0f)
```